pkg/ipamclaimswebhook/podmutator.go

```diff
@@ -204,3 +170,117 @@ func podWithUpdatedSelectionElements(pod *corev1.Pod, networks []v1.NetworkSelec
 	newPod.Annotations[v1.NetworkAttachmentAnnot] = string(newNets)
 	return newPod, nil
 }
+
+func ensureIPAMClaimRefAtNetworkSelectionElements(ctx context.Context,
```

Comment:
I really dislike the fact this returns 3 things.

And the fact we use this `response` parameter just to indicate something that IMO looks like an error is very puzzling to me.

Reply:
I don't like it either, but I was trying to maintain the original behaviour where is returning `NAD not found, will hang on scheduler` if there is no nad for the NSE, we can just return `nil, nil` but logging it and the webhook response will be "carry on"

Reply:
I suggest you return a new error type like "NoNadNotOurProblem" (I'm bad with names but you get my drift) and expect that case outside of `ensureIPAMClaimRefAtNetworkSelectionElements`, returning the AdmissionAllowed

Reply:
> I suggest you return a new error type like "NoNadIgnoreReqest" and expect that case outside of `ensureIPAMClaimRefAtNetworkSelectionElements`, returning the AdmissionAllowed
> 
> ```
> var ErrNoNadIgnoreRequest = errors.New("NAD not found, will hang on scheduler")
> ```
> 
> on hasChangedNetworkSelectionElements function:
> 
> ```
> if k8serrors.IsNotFound(err) {
>   return ErrNoNadIgnoreRequest
> }
> ```
> 
> then outside of it:
> 
> ```
> hasChangedNetworkSelectionElements, err :=
>     ensureIPAMClaimRefAtNetworkSelectionElements(ctx, a.Client, pod, vmName, networkSelectionElements)
> if err != nil {
>     if errors.Is(err, ErrNoNadIgnoreRequest) {
>         return admission.Allowed(ErrNoNadIgnoreRequest.Error())
>     }
>     return admission.Errored(http.StatusInternalServerError, err)
> }
> ```

Puff I think this is overkill just to return something at the webhook success, I want to know if returning that specific string is really necessary or logging it is enough.

Reply:
hm, returning bool, error also makes me frown...

Reply:
ok, simplified since we don't care about th emssage at webhook.Allowed

Reply:
no sure you saw it and chose to ignore it, or just didn't see it. 

> hm, returning bool, error also makes me frown...

Can we try to improve this API ?

Reply:
What the issue with that ? is a function that try to update a map and return if the map was updated or not.

Reply:
you can do the exact same thing with just error, or returning the map and letting the client understand if it changed or not. 

if your function does "update a map and return if the map was updated or not" your function is doing 2 things. The function should do one. Update the map. 

We can improve it later on if this is the only item under question.

Reply:
openned an issue to simplify things later on https://github.com/kubevirt/ipam-extensions/issues/63